sasdata/dataloader/filereader.py

```diff
@@ -74,62 +76,71 @@ def __init__(self):
         self.current_datainfo = None
         # File path sent to reader
         self.filepath = None
+        # File extension sent to reader
+        self.extension = None
         # Open file handle
         self.f_open = None
+        # Open string file handle
+        self.string_open = None
+        # Open HDF file handle
+        self.hdf_open = None
 
-    def read(self, filepath: str) -> List[Union[Data1D, Data2D]]:
+    def read(self, filepath: str, file_handler: Optional[CustomFileOpen] = None) -> List[Union[Data1D, Data2D]]:
         """
         Basic file reader
 
-        :param filepath: The full or relative path to a file to be loaded
+        :param filepath: The string representation of the path to a file to be loaded. This can be a URI or a local file
+        :param raw_file: An open TextIO or Binary IO file handle to be read.
+        :param hdf5_file: An open h5py.File file handle, which is also an h5py Group.
         """
         self.filepath = filepath
-        if os.path.isfile(filepath):
-            basename, extension = os.path.splitext(os.path.basename(filepath))
-            self.extension = extension.lower()
-            # If the file type is not allowed, return nothing
-            if self.extension in self.ext or self.allow_all:
-                # Try to load the file, but raise an error if unable to.
-                try:
-                    with open(filepath, 'rb') as self.f_open:
-                        self.get_file_contents()
-                except DataReaderException as e:
-                    self.handle_error_message(str(e))
-                except FileContentsException as e:
-                    raise
-                except OSError as e:
-                    # If the file cannot be opened
-                    msg = "Unable to open file: {}\n".format(filepath)
-                    msg += str(e)
-                    self.handle_error_message(msg)
-                except Exception as e:
-                    self.handle_error_message(str(e))
-                finally:
-                    if any(filepath.lower().endswith(ext) for ext in
-                           self.deprecated_extensions):
-                        self.handle_error_message(DEPRECATION_MESSAGE)
-                    if len(self.output) > 0:
-                        # Sort the data that's been loaded
-                        self.convert_data_units()
-                        self.sort_data()
+        if isinstance(file_handler, CustomFileOpen):
+            self.f_open = file_handler.fd
+            self.hdf_open = file_handler.h5_fd
+            self.string_open = file_handler.string_fd
         else:
-            msg = "Unable to find file at: {}\n".format(filepath)
-            msg += "Please check your file path and try again."
+            # If not called by a context manager, create one locally that calls read() again. Return early
+            return self._read_with_local_context_manager(filepath)
```

Comment:
This gets confusing to debug.

I suggest either creating a new reader class than accepts a byte stream and replace the existing with a shim to call the new reader to preserve the interface, or perhaps a new method that accepts a byte stream, with the existing read method opening the byte stream and returning a call to the new method.